File: app/src/main/java/com/drdisagree/iconify/ui/activity/QsIconLabel.java
package com.drdisagree.iconify.ui.activity;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.drdisagree.iconify.Iconify;
import com.drdisagree.iconify.R;
import com.drdisagree.iconify.config.PrefConfig;
import com.drdisagree.iconify.overlaymanager.IconSizeManager;
import com.drdisagree.iconify.overlaymanager.QSMoveIconManager;
import com.drdisagree.iconify.overlaymanager.QSTextSizeManager;
import com.drdisagree.iconify.utils.OverlayUtil;
import com.google.android.material.appbar.CollapsingToolbarLayout;

import java.util.Objects;

public class QsIconLabel extends AppCompatActivity {

    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_qs_icon_label);

        // Header
        CollapsingToolbarLayout collapsing_toolbar = findViewById(R.id.collapsing_toolbar);
        collapsing_toolbar.setTitle(getResources().getString(R.string.activity_title_qs_icon_label));
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        // Text Size
        SeekBar text_size = findViewById(R.id.text_size);
        TextView text_size_output = findViewById(R.id.text_size_output);

        text_size.setPadding(0, 0, 0, 0);
        final int[] finalTextSize = {4};

        if (!PrefConfig.loadPrefSettings("qsTextSize").equals("null")) {
            if ((Integer.parseInt(PrefConfig.loadPrefSettings("qsTextSize")) + 10) == 14)
                text_size_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsTextSize")) + 10) + "sp " + getResources().getString(R.string.opt_default));
            else
                text_size_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsTextSize")) + 10) + "sp");
            finalTextSize[0] = Integer.parseInt(PrefConfig.loadPrefSettings("qsTextSize"));
            text_size.setProgress(finalTextSize[0]);
        } else
            text_size_output.setText(getResources().getString(R.string.opt_selected) + " 14sp " + getResources().getString(R.string.opt_default));

        text_size.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

                finalTextSize[0] = progress;
                if (progress + 10 == 14)
                    text_size_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 10) + "sp " + getResources().getString(R.string.opt_default));
                else
                    text_size_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 10) + "sp");

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                PrefConfig.savePrefSettings("qsTextSize", String.valueOf(finalTextSize[0]));
                PrefConfig.savePrefBool("fabricatedqsTextSize", true);

                Runnable runnable = () -> QSTextSizeManager.install_pack(finalTextSize[0]);
                Thread thread = new Thread(runnable);
                thread.start();

                Toast.makeText(Iconify.getAppContext(), (finalTextSize[0] + 10 + "sp " + getResources().getString(R.string.toast_applied)), Toast.LENGTH_SHORT).show();
            }
        });

        // Icon Size

        SeekBar icon_size = findViewById(R.id.icon_size);
        TextView icon_size_output = findViewById(R.id.icon_size_output);

        icon_size.setPadding(0, 0, 0, 0);
        final int[] finalIconSize = {10};

        if (!PrefConfig.loadPrefSettings("qsIconSize").equals("null")) {
            if ((Integer.parseInt(PrefConfig.loadPrefSettings("qsIconSize")) + 10) == 20)
                icon_size_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsIconSize")) + 10) + "dp " + getResources().getString(R.string.opt_default));
            else
                icon_size_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsIconSize")) + 10) + "dp");
            finalIconSize[0] = Integer.parseInt(PrefConfig.loadPrefSettings("qsIconSize"));
            icon_size.setProgress(finalIconSize[0]);
        } else
            icon_size_output.setText(getResources().getString(R.string.opt_selected) + " 20dp " + getResources().getString(R.string.opt_default));

        icon_size.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

                finalIconSize[0] = progress;
                if (progress + 10 == 20)
                    icon_size_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 10) + "dp " + getResources().getString(R.string.opt_default));
                else
                    icon_size_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 10) + "dp");
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                PrefConfig.savePrefSettings("qsIconSize", String.valueOf(finalIconSize[0]));
                PrefConfig.savePrefBool("fabricatedqsIconSize", true);

                Runnable runnable = () -> IconSizeManager.install_pack(finalIconSize[0]);
                Thread thread = new Thread(runnable);
                thread.start();

                Toast.makeText(Iconify.getAppContext(), (finalIconSize[0] + 10 + "dp " + getResources().getString(R.string.toast_applied)), Toast.LENGTH_SHORT).show();
            }
        });

        // Hide text size if hide label is enabled

        LinearLayout text_size_container = findViewById(R.id.text_size_container);
        LinearLayout icon_size_container = findViewById(R.id.icon_size_container);
        View text_size_divider = findViewById(R.id.text_size_divider);

        if (PrefConfig.loadPrefBool("IconifyComponentQSHL.overlay")) {
            text_size_container.setVisibility(View.GONE);
            text_size_divider.setVisibility(View.GONE);
        }

        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch label_white = findViewById(R.id.label_white);
        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch label_whiteV2 = findViewById(R.id.label_whiteV2);
        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch label_systemInverse = findViewById(R.id.label_systemInverse);
        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch label_systemInverseV2 = findViewById(R.id.label_systemInverseV2);
        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch label_fixtexta13 = findViewById(R.id.label_fixtexta13);

        label_white.setChecked(PrefConfig.loadPrefBool("IconifyComponentQST1.overlay"));

        label_white.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                label_whiteV2.setChecked(false);
                label_systemInverse.setChecked(false);
                label_systemInverseV2.setChecked(false);
                label_fixtexta13.setChecked(false);
                OverlayUtil.disableOverlay("IconifyComponentQST2.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST3.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST4.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST5.overlay");

                OverlayUtil.enableOverlay("IconifyComponentQST1.overlay");
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQST1.overlay");
            }
        });

        label_whiteV2.setChecked(PrefConfig.loadPrefBool("IconifyComponentQST2.overlay"));

        label_whiteV2.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                label_white.setChecked(false);
                label_systemInverse.setChecked(false);
                label_systemInverseV2.setChecked(false);
                label_fixtexta13.setChecked(false);
                OverlayUtil.disableOverlay("IconifyComponentQST1.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST3.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST4.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST5.overlay");

                OverlayUtil.enableOverlay("IconifyComponentQST2.overlay");
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQST2.overlay");
            }
        });

        label_systemInverse.setChecked(PrefConfig.loadPrefBool("IconifyComponentQST3.overlay"));

        label_systemInverse.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                label_white.setChecked(false);
                label_whiteV2.setChecked(false);
                label_systemInverseV2.setChecked(false);
                label_fixtexta13.setChecked(false);
                OverlayUtil.disableOverlay("IconifyComponentQST1.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST2.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST4.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST5.overlay");

                OverlayUtil.enableOverlay("IconifyComponentQST3.overlay");
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQST3.overlay");
            }
        });

        label_systemInverseV2.setChecked(PrefConfig.loadPrefBool("IconifyComponentQST4.overlay"));

        label_systemInverseV2.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                label_white.setChecked(false);
                label_whiteV2.setChecked(false);
                label_systemInverse.setChecked(false);
                label_fixtexta13.setChecked(false);
                OverlayUtil.disableOverlay("IconifyComponentQST1.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST2.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST3.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST5.overlay");

                OverlayUtil.enableOverlay("IconifyComponentQST4.overlay");
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQST4.overlay");
            }
        });

        label_fixtexta13.setChecked(PrefConfig.loadPrefBool("IconifyComponentQST5.overlay"));

        label_fixtexta13.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                label_white.setChecked(false);
                label_whiteV2.setChecked(false);
                label_systemInverse.setChecked(false);
                label_systemInverseV2.setChecked(false);

                OverlayUtil.disableOverlay("IconifyComponentQST1.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST2.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST3.overlay");
                OverlayUtil.disableOverlay("IconifyComponentQST4.overlay");

                OverlayUtil.enableOverlay("IconifyComponentQST5.overlay");
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQST5.overlay");
            }
        });

        // Hide Label

        @SuppressLint("UseSwitchCompatOrMaterialCode") Switch hide_label = findViewById(R.id.hide_label);

        hide_label.setChecked(PrefConfig.loadPrefBool("IconifyComponentQSHL.overlay"));

        hide_label.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) {
                OverlayUtil.enableOverlay("IconifyComponentQSHL.overlay");

                text_size_container.setVisibility(View.GONE);
                text_size_divider.setVisibility(View.GONE);
            } else {
                OverlayUtil.disableOverlay("IconifyComponentQSHL.overlay");

                text_size_container.setVisibility(View.VISIBLE);
                text_size_divider.setVisibility(View.VISIBLE);
            }
        });

        // Move Icon

        SeekBar move_icon = findViewById(R.id.move_icon);
        TextView move_icon_output = findViewById(R.id.move_icon_output);

        move_icon.setPadding(0, 0, 0, 0);
        final int[] finalMoveIcon = {15};

        if (!PrefConfig.loadPrefSettings("qsMoveIcon").equals("null")) {
            if ((Integer.parseInt(PrefConfig.loadPrefSettings("qsMoveIcon")) + 1) == 16)
                move_icon_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsMoveIcon")) + 1) + "dp " + getResources().getString(R.string.opt_default));
            else
                move_icon_output.setText(getResources().getString(R.string.opt_selected) + (Integer.parseInt(PrefConfig.loadPrefSettings("qsMoveIcon")) + 1) + "dp");
            finalMoveIcon[0] = Integer.parseInt(PrefConfig.loadPrefSettings("qsMoveIcon"));
            move_icon.setProgress(finalMoveIcon[0]);
        } else
            move_icon_output.setText(getResources().getString(R.string.opt_selected) + " 16dp " + getResources().getString(R.string.opt_default));

        move_icon.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {

                finalMoveIcon[0] = progress;
                if (progress + 1 == 16)
                    move_icon_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 1) + "dp " + getResources().getString(R.string.opt_default));
                else
                    move_icon_output.setText(getResources().getString(R.string.opt_selected) + ' ' + (progress + 1) + "dp");
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                PrefConfig.savePrefSettings("qsMoveIcon", String.valueOf(finalMoveIcon[0]));
                PrefConfig.savePrefBool("fabricatedqsMoveIcon", true);

                Runnable runnable = () -> QSMoveIconManager.install_pack(finalMoveIcon[0]);
                Thread thread = new Thread(runnable);
                thread.start();

                Toast.makeText(Iconify.getAppContext(), (finalMoveIcon[0] + 1 + "dp " + getResources().getString(R.string.toast_applied)), Toast.LENGTH_SHORT).show();
            }
        });

    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: app/src/main/java/com/drdisagree/iconify/ui/fragments/tweaks/QsIconLabel.kt
package com.drdisagree.iconify.ui.fragments.tweaks

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.Toast
import com.drdisagree.iconify.R
import com.drdisagree.iconify.common.Const.SWITCH_ANIMATION_DELAY
import com.drdisagree.iconify.common.Const.SYSTEMUI_PACKAGE
import com.drdisagree.iconify.common.Dynamic.isAtleastA14
import com.drdisagree.iconify.common.Preferences.QS_HIDE_LABEL_SWITCH
import com.drdisagree.iconify.common.Preferences.QS_TEXT_COLOR_VARIANT
import com.drdisagree.iconify.common.Preferences.QS_TEXT_COLOR_VARIANT_NORMAL
import com.drdisagree.iconify.common.Preferences.QS_TEXT_COLOR_VARIANT_PIXEL
import com.drdisagree.iconify.common.References.FABRICATED_QS_ICON_SIZE
import com.drdisagree.iconify.common.References.FABRICATED_QS_MOVE_ICON
import com.drdisagree.iconify.common.References.FABRICATED_QS_TEXT_SIZE
import com.drdisagree.iconify.config.RPrefs.getBoolean
import com.drdisagree.iconify.config.RPrefs.getString
import com.drdisagree.iconify.config.RPrefs.putBoolean
import com.drdisagree.iconify.config.RPrefs.putString
import com.drdisagree.iconify.databinding.FragmentQsIconLabelBinding
import com.drdisagree.iconify.ui.base.BaseFragment
import com.drdisagree.iconify.ui.utils.ViewHelper.setHeader
import com.drdisagree.iconify.utils.SystemUtils.hasStoragePermission
import com.drdisagree.iconify.utils.SystemUtils.requestStoragePermission
import com.drdisagree.iconify.utils.overlay.OverlayUtils
import com.drdisagree.iconify.utils.overlay.OverlayUtils.changeOverlayState
import com.drdisagree.iconify.utils.overlay.OverlayUtils.disableOverlays
import com.drdisagree.iconify.utils.overlay.OverlayUtils.enableOverlay
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceEntry
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.buildOverlayWithResource
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager.removeResourceFromOverlay
import com.google.android.material.button.MaterialButtonToggleGroup
import com.google.android.material.slider.Slider
import java.util.concurrent.atomic.AtomicBoolean

class QsIconLabel : BaseFragment() {

    private lateinit var binding: FragmentQsIconLabelBinding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentQsIconLabelBinding.inflate(inflater, container, false)
        val view: View = binding.getRoot()

        // Header
        setHeader(
            requireContext(),
            getParentFragmentManager(),
            binding.header.toolbar,
            R.string.activity_title_qs_icon_label
        )

        // Text Size
        val finalTextSize = intArrayOf(14)
        if (getString(FABRICATED_QS_TEXT_SIZE) != null) {
            finalTextSize[0] = getString(FABRICATED_QS_TEXT_SIZE)!!.toInt()
            binding.textSize.sliderValue = finalTextSize[0]
        }

        // Reset button
        binding.textSize.setResetClickListener {
            finalTextSize[0] = 14
            putString(FABRICATED_QS_TEXT_SIZE, finalTextSize[0].toString())

            removeResourceFromOverlay(
                requireContext(),
                ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_tile_text_size")
            )

            true
        }
        binding.textSize.setOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalTextSize[0] = slider.value.toInt()
                putString(FABRICATED_QS_TEXT_SIZE, finalTextSize[0].toString())

                buildOverlayWithResource(
                    requireContext(),
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "qs_tile_text_size",
                        finalTextSize[0].toString() + "sp"
                    )
                )
            }
        })

        // Icon Size
        val finalIconSize = intArrayOf(20)
        if (getString(FABRICATED_QS_ICON_SIZE) != null) {
            finalIconSize[0] = getString(FABRICATED_QS_ICON_SIZE)!!.toInt()
            binding.iconSize.sliderValue = finalIconSize[0]
        }

        // Reset button
        binding.iconSize.setResetClickListener {
            finalIconSize[0] = 20
            putString(FABRICATED_QS_ICON_SIZE, finalIconSize[0].toString())

            removeResourceFromOverlay(
                requireContext(),
                ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_icon_size")
            )

            true
        }
        binding.iconSize.setOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalIconSize[0] = slider.value.toInt()
                putString(FABRICATED_QS_ICON_SIZE, finalIconSize[0].toString())

                buildOverlayWithResource(
                    requireContext(),
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "qs_icon_size",
                        finalIconSize[0].toString() + "sp"
                    )
                )
            }
        })

        // Hide text size if hide label is enabled
        if (getBoolean(QS_HIDE_LABEL_SWITCH, false)) {
            binding.textSize.visibility = View.GONE
        }

        // QS Text Color
        if (isNormalVariantActive) {
            binding.toggleButtonTextColor.check(R.id.textColorNormal)
        } else if (isPixelVariantActive) {
            binding.toggleButtonTextColor.check(R.id.textColorPixel)
        }

        selectedVariant =
            if (binding.toggleButtonTextColor.checkedButtonId == R.id.textColorNormal) {
                QS_TEXT_COLOR_VARIANT_NORMAL
            } else {
                QS_TEXT_COLOR_VARIANT_PIXEL
            }

        binding.toggleButtonTextColor.addOnButtonCheckedListener { _: MaterialButtonToggleGroup?, checkedId: Int, _: Boolean ->
            if (checkedId == R.id.textColorNormal) {
                if (selectedVariant != QS_TEXT_COLOR_VARIANT_NORMAL) {
                    putString(QS_TEXT_COLOR_VARIANT, QS_TEXT_COLOR_VARIANT_NORMAL)
                    selectedVariant = QS_TEXT_COLOR_VARIANT_NORMAL

                    if (getBoolean(QSPT1_OVERLAY)) changeOverlayState(
                        QSPT1_OVERLAY,
                        false,
                        QSNT1_OVERLAY,
                        true
                    )

                    if (getBoolean(QSPT2_OVERLAY)) changeOverlayState(
                        QSPT2_OVERLAY,
                        false,
                        QSNT2_OVERLAY,
                        true
                    )

                    if (getBoolean(QSPT3_OVERLAY)) changeOverlayState(
                        QSPT3_OVERLAY,
                        false,
                        QSNT3_OVERLAY,
                        true
                    )

                    if (getBoolean(QSPT4_OVERLAY)) changeOverlayState(
                        QSPT4_OVERLAY,
                        false,
                        QSNT4_OVERLAY,
                        true
                    )

                    handleCommonOverlay()
                }
            } else if (checkedId == R.id.textColorPixel) {
                if (selectedVariant != QS_TEXT_COLOR_VARIANT_PIXEL) {
                    putString(QS_TEXT_COLOR_VARIANT, QS_TEXT_COLOR_VARIANT_PIXEL)
                    selectedVariant = QS_TEXT_COLOR_VARIANT_PIXEL

                    if (getBoolean(QSNT1_OVERLAY)) changeOverlayState(
                        QSNT1_OVERLAY,
                        false,
                        QSPT1_OVERLAY,
                        true
                    )

                    if (getBoolean(QSNT2_OVERLAY)) changeOverlayState(
                        QSNT2_OVERLAY,
                        false,
                        QSPT2_OVERLAY,
                        true
                    )

                    if (getBoolean(QSNT3_OVERLAY)) changeOverlayState(
                        QSNT3_OVERLAY,
                        false,
                        QSPT3_OVERLAY,
                        true
                    )

                    if (getBoolean(QSNT4_OVERLAY)) changeOverlayState(
                        QSNT4_OVERLAY,
                        false,
                        QSPT4_OVERLAY,
                        true
                    )

                    handleCommonOverlay()
                }
            }
        }

        binding.labelWhite.isSwitchChecked =
            getBoolean(replaceVariant("IconifyComponentQST1.overlay"))
        binding.labelWhite.setSwitchChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked && isAtleastA14) {
                Toast.makeText(
                    requireContext(),
                    R.string.toast_use_from_xposed_menu,
                    Toast.LENGTH_SHORT
                ).show()

                return@setSwitchChangeListener
            }

            Handler(Looper.getMainLooper()).postDelayed({
                if (isChecked) {
                    binding.labelWhiteV2.isSwitchChecked = false
                    binding.labelSystem.isSwitchChecked = false
                    binding.labelSystemV2.isSwitchChecked = false
                    binding.labelFixTextColor.isSwitchChecked = false

                    disableOverlays(
                        *replaceVariant(
                            "IconifyComponentQST2.overlay",
                            "IconifyComponentQST3.overlay",
                            "IconifyComponentQST4.overlay",
                            "IconifyComponentQST5.overlay"
                        )
                    )

                    enableOverlay(replaceVariant("IconifyComponentQST1.overlay"))
                } else {
                    OverlayUtils.disableOverlay(replaceVariant("IconifyComponentQST1.overlay"))
                }

                handleCommonOverlay()
            }, SWITCH_ANIMATION_DELAY)
        }

        binding.labelWhiteV2.isSwitchChecked =
            getBoolean(replaceVariant("IconifyComponentQST2.overlay"))
        binding.labelWhiteV2.setSwitchChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked && isAtleastA14) {
                Toast.makeText(
                    requireContext(),
                    R.string.toast_use_from_xposed_menu,
                    Toast.LENGTH_SHORT
                ).show()

                return@setSwitchChangeListener
            }

            Handler(Looper.getMainLooper()).postDelayed({
                if (isChecked) {
                    binding.labelWhite.isSwitchChecked = false
                    binding.labelSystem.isSwitchChecked = false
                    binding.labelSystemV2.isSwitchChecked = false
                    binding.labelFixTextColor.isSwitchChecked = false

                    disableOverlays(
                        *replaceVariant(
                            "IconifyComponentQST1.overlay",
                            "IconifyComponentQST3.overlay",
                            "IconifyComponentQST4.overlay",
                            "IconifyComponentQST5.overlay"
                        )
                    )

                    enableOverlay(replaceVariant("IconifyComponentQST2.overlay"))
                } else {
                    OverlayUtils.disableOverlay(replaceVariant("IconifyComponentQST2.overlay"))
                }

                handleCommonOverlay()
            }, SWITCH_ANIMATION_DELAY)
        }

        binding.labelSystem.isSwitchChecked =
            getBoolean(replaceVariant("IconifyComponentQST3.overlay"))
        binding.labelSystem.setSwitchChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked && isAtleastA14) {
                Toast.makeText(
                    requireContext(),
                    R.string.toast_use_from_xposed_menu,
                    Toast.LENGTH_SHORT
                ).show()

                return@setSwitchChangeListener
            }

            Handler(Looper.getMainLooper()).postDelayed({
                if (isChecked) {
                    binding.labelWhite.isSwitchChecked = false
                    binding.labelWhiteV2.isSwitchChecked = false
                    binding.labelSystemV2.isSwitchChecked = false
                    binding.labelFixTextColor.isSwitchChecked = false

                    disableOverlays(
                        *replaceVariant(
                            "IconifyComponentQST1.overlay",
                            "IconifyComponentQST2.overlay",
                            "IconifyComponentQST4.overlay",
                            "IconifyComponentQST5.overlay"
                        )
                    )

                    enableOverlay(replaceVariant("IconifyComponentQST3.overlay"))
                } else {
                    OverlayUtils.disableOverlay(replaceVariant("IconifyComponentQST3.overlay"))
                }

                handleCommonOverlay()
            }, SWITCH_ANIMATION_DELAY)
        }

        binding.labelSystemV2.isSwitchChecked =
            getBoolean(replaceVariant("IconifyComponentQST4.overlay"))
        binding.labelSystemV2.setSwitchChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked && isAtleastA14) {
                Toast.makeText(
                    requireContext(),
                    R.string.toast_use_from_xposed_menu,
                    Toast.LENGTH_SHORT
                ).show()

                return@setSwitchChangeListener
            }
            Handler(Looper.getMainLooper()).postDelayed({
                if (isChecked) {
                    binding.labelWhite.isSwitchChecked = false
                    binding.labelWhiteV2.isSwitchChecked = false
                    binding.labelSystem.isSwitchChecked = false
                    binding.labelFixTextColor.isSwitchChecked = false

                    disableOverlays(
                        *replaceVariant(
                            "IconifyComponentQST1.overlay",
                            "IconifyComponentQST2.overlay",
                            "IconifyComponentQST3.overlay",
                            "IconifyComponentQST5.overlay"
                        )
                    )

                    enableOverlay(replaceVariant("IconifyComponentQST4.overlay"))
                } else {
                    OverlayUtils.disableOverlay(replaceVariant("IconifyComponentQST4.overlay"))
                }

                handleCommonOverlay()
            }, SWITCH_ANIMATION_DELAY)
        }

        binding.labelFixTextColor.isSwitchChecked =
            getBoolean("IconifyComponentQST5.overlay")
        binding.labelFixTextColor.setSwitchChangeListener { _: CompoundButton?, isChecked: Boolean ->
            if (isChecked && isAtleastA14) {
                Toast.makeText(
                    requireContext(),
                    R.string.toast_use_from_xposed_menu,
                    Toast.LENGTH_SHORT
                ).show()

                return@setSwitchChangeListener
            }

            Handler(Looper.getMainLooper()).postDelayed({
                if (isChecked) {
                    binding.labelWhite.isSwitchChecked = false
                    binding.labelWhiteV2.isSwitchChecked = false
                    binding.labelSystem.isSwitchChecked = false
                    binding.labelSystemV2.isSwitchChecked = false

                    disableOverlays(
                        *replaceVariant(
                            "IconifyComponentQST1.overlay",
                            "IconifyComponentQST2.overlay",
                            "IconifyComponentQST3.overlay",
                            "IconifyComponentQST4.overlay"
                        )
                    )

                    enableOverlay(replaceVariant("IconifyComponentQST5.overlay"))
                } else {
                    OverlayUtils.disableOverlay(replaceVariant("IconifyComponentQST5.overlay"))
                }

                handleCommonOverlay()
            }, SWITCH_ANIMATION_DELAY)
        }

        // Hide Label
        val isHideLabelContainerClicked = AtomicBoolean(false)

        binding.hideLabel.isSwitchChecked = getBoolean(QS_HIDE_LABEL_SWITCH, false)
        binding.hideLabel.setSwitchChangeListener { buttonView: CompoundButton, isChecked: Boolean ->
            if (!hasStoragePermission()) {
                isHideLabelContainerClicked.set(false)
                requestStoragePermission(requireContext())
                binding.hideLabel.isSwitchChecked = !isChecked
            } else if (buttonView.isPressed || isHideLabelContainerClicked.get()) {
                isHideLabelContainerClicked.set(false)
                putBoolean(QS_HIDE_LABEL_SWITCH, isChecked)

                if (isChecked) {
                    buildOverlayWithResource(
                        requireContext(),
                        ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_tile_text_size", "0sp"),
                        ResourceEntry(
                            SYSTEMUI_PACKAGE,
                            "dimen",
                            "qs_label_container_margin",
                            "-120dp"
                        )
                    )

                    binding.textSize.visibility = View.GONE
                } else {
                    removeResourceFromOverlay(
                        requireContext(),
                        ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_tile_text_size"),
                        ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_label_container_margin")
                    )

                    binding.textSize.visibility = View.VISIBLE
                }
            }
        }
        binding.hideLabel.setBeforeSwitchChangeListener {
            isHideLabelContainerClicked.set(true)
        }

        // Move Icon
        val finalMoveIcon = intArrayOf(16)
        if (getString(FABRICATED_QS_MOVE_ICON) != null) {
            finalMoveIcon[0] = getString(FABRICATED_QS_MOVE_ICON)!!.toInt()
            binding.moveIcon.sliderValue = finalMoveIcon[0]
        }

        // Reset button
        binding.moveIcon.setResetClickListener {
            finalMoveIcon[0] = 16
            putString(FABRICATED_QS_MOVE_ICON, finalMoveIcon[0].toString())

            removeResourceFromOverlay(
                requireContext(),
                ResourceEntry(SYSTEMUI_PACKAGE, "dimen", "qs_tile_start_padding")
            )

            true
        }
        binding.moveIcon.setOnSliderTouchListener(object : Slider.OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                finalMoveIcon[0] = slider.value.toInt()
                putString(FABRICATED_QS_MOVE_ICON, finalMoveIcon[0].toString())

                buildOverlayWithResource(
                    requireContext(),
                    ResourceEntry(
                        SYSTEMUI_PACKAGE,
                        "dimen",
                        "qs_tile_start_padding",
                        finalMoveIcon[0].toString() + "dp"
                    )
                )
            }
        })

        return view
    }

    private fun replaceVariant(vararg args: String): Array<String?> {
        val newArgs = arrayOfNulls<String>(args.size)

        for (i in args.indices) {
            if (args[i].contains("QST5")) continue

            newArgs[i] = args[i].replace(
                "QST",
                if (selectedVariant == QS_TEXT_COLOR_VARIANT_NORMAL) "QSNT" else "QSPT"
            )
        }

        return newArgs
    }

    private fun replaceVariant(arg: String): String {
        return if (arg.contains("QST5")) arg else arg.replace(
            "QST",
            if (selectedVariant == QS_TEXT_COLOR_VARIANT_NORMAL) "QSNT" else "QSPT"
        )
    }

    private val isNormalVariantActive: Boolean
        get() = getBoolean(QSNT1_OVERLAY) ||
                getBoolean(QSNT2_OVERLAY) ||
                getBoolean(QSNT3_OVERLAY) ||
                getBoolean(QSNT4_OVERLAY) || getString(QS_TEXT_COLOR_VARIANT) == QS_TEXT_COLOR_VARIANT_NORMAL

    private val isPixelVariantActive: Boolean
        get() = getBoolean(QSPT1_OVERLAY) ||
                getBoolean(QSPT2_OVERLAY) ||
                getBoolean(QSPT3_OVERLAY) ||
                getBoolean(QSPT4_OVERLAY) || getString(QS_TEXT_COLOR_VARIANT) == QS_TEXT_COLOR_VARIANT_PIXEL

    private fun handleCommonOverlay() {
        changeOverlayState(
            QSNPT_OVERLAY,
            getBoolean(QSNT1_OVERLAY) ||
                    getBoolean(QSNT2_OVERLAY) ||
                    getBoolean(QSNT3_OVERLAY) ||
                    getBoolean(QSNT4_OVERLAY) ||
                    getBoolean(QSPT1_OVERLAY) ||
                    getBoolean(QSPT2_OVERLAY) ||
                    getBoolean(QSPT3_OVERLAY) ||
                    getBoolean(QSPT4_OVERLAY)
        )
    }

    companion object {
        private var selectedVariant: String? = null
        const val QSNPT_OVERLAY = "IconifyComponentQSNPT.overlay"
        const val QSNT1_OVERLAY = "IconifyComponentQSNT1.overlay"
        const val QSNT2_OVERLAY = "IconifyComponentQSNT2.overlay"
        const val QSNT3_OVERLAY = "IconifyComponentQSNT3.overlay"
        const val QSNT4_OVERLAY = "IconifyComponentQSNT4.overlay"
        const val QSPT1_OVERLAY = "IconifyComponentQSPT1.overlay"
        const val QSPT2_OVERLAY = "IconifyComponentQSPT2.overlay"
        const val QSPT3_OVERLAY = "IconifyComponentQSPT3.overlay"
        const val QSPT4_OVERLAY = "IconifyComponentQSPT4.overlay"
    }
}